docs: adds focus management common pattern docs (#217)
primer/design: content/ui-patterns/focus-management.mdx

---
title: Focus management
---

## Overview

> Managing focus within a page or application is essential for users to successfully navigate, complete actions, and understand where they are within a web page or application. A focus indicator is usually seen visually as a blue outline around an interactive element when that element has focus.

Reviewer: ```suggestion
Managing focus within a page or application is essential for users to successfully navigate, complete actions, and understand where they are. A focus indicator is usually seen visually as a blue outline around an interactive element when that element has focus.
```

## Why?

Not all users are able to use a mouse to navigate a web page or application. Keyboard-only and screen reader users rely on navigating and using a web page with only a keyboard, so they will be unable to perform certain actions if they cannot navigate to all elements.

To meet [WCAG Guideline 2.1 (Keyboard Accessible)](https://www.w3.org/TR/WCAG21/#keyboard-accessible), all interactive elements need to be accessed and activated by a keyboard alone.

## How to test focus management

> To verify proper focus management, try using the feature or page with only a keyboard. Use the Tab key to navigate forward and Shift + Tab to move backwards.

Reviewer 2: ```suggestion
To verify proper focus management, try using the feature or page with only a keyboard. Use the Tab key to navigate forward and Shift + Tab to move backwards. Some elements will use the arrow keys for keyboard navigation, such as toolbars and dropdowns.
```

> Things to keep in mind:

Reviewer: ```suggestion
While using the feature, you should make sure that:
```

> - You can navigate to every interactive element (i.e. buttons, form fields, links)
> - You see a visual focus indicator for every element (i.e. a blue outline)
> - The tabbing order is logical based on the page structure and content
> - You are not trapped anywhere, and focus is never lost (i.e. when deleting a comment, focus is returned to a logical place)
> - If you can trigger a new element (i.e. a modal), focus is moved to that modal when opened, and returned to the page when closed

Reviewer: ```suggestion
- You can navigate to every interactive element (for example: buttons, form fields, links)
- You see a visual focus indicator for every element (for example: a blue outline)
- The tabbing order is logical based on the page structure and content
- You are not trapped anywhere, and focus is never lost (for example: when deleting a comment, focus is returned to a logical place)
- If you can trigger a new element (for example: a modal), focus is moved to that modal when opened, and returned to the page when closed
```

> ## For designers

Reviewer: I'm thinking the "For designers" and "For engineers" headings should actually be h3 and there should be a h2 heading above them with a short intro of _what_ is for designers? Are the guidelines? Something else?

> Notate how the focus should move through various states of a feature, if applicable, for the engineer. Include where the focus should be when: an item is deleted, an item is added, an action triggers a page refresh, a modal is opened, a modal is closed, etc.

Reviewer: ```suggestion
In your designs, annotate how the focus should move through various states of a feature, if applicable, for the user. Include where the focus should be when: an item is deleted, an item is added, an action triggers a page refresh, a modal is opened, a modal is closed, etc.
```

### Adding content

When adding content to the page, focus should move to the first item that is added, unless a more logical spot is determined. For example: if there is a “Load more” button that reveals additional comments when pressed, the focus should move to the first comment that is newly added.

### Removing content

When removing content, focus should return to a logical spot. By default, if you remove an item and don't tell the focus where to go next, it will return to the `body` element and the user will need to navigate through the entire page again to get to where they were. For example, if deleting a comment within a thread, move the focus to either a “write comment” input if one exists, or the menu of the next comment. If there is no other comment, use the previous comment. This is just a suggestion, use best judgement when deciding what will be logical for a user.

> ## Additional Resources

Reviewer: ```suggestion
## Additional resources
```

> ### Good Focus Management

Reviewer: ```suggestion
### Good focus management
```

- [Improve Accessibility in Your React App By Managing Focus in Mutable Content](https://medium.com/swlh/improve-accessibility-in-your-react-app-by-managing-focus-in-mutable-content-4ddf4ed92186)
- [Focus management and inert](https://css-tricks.com/focus-management-and-inert/)
- [Removing Headaches from Focus Management](https://developers.google.com/web/updates/2016/03/focus-start-point)

> ### Related WCAG Guidelines & Success Criteria:

Reviewer: ```suggestion
### Related WCAG guidelines and success criteria
```
